internal/component/mimir/rules/kubernetes/events.go

```diff
@@ -157,38 +201,71 @@ func convertCRDRuleGroupToRuleGroup(crd promv1.PrometheusRuleSpec) ([]rulefmt.Ru
 	return groups.Groups, nil
 }
 
-func (c *Component) applyChanges(ctx context.Context, namespace string, diffs []kubernetes.RuleGroupDiff) error {
+func (e *eventProcessor) applyChanges(ctx context.Context, namespace string, diffs []kubernetes.RuleGroupDiff) error {
 	if len(diffs) == 0 {
 		return nil
 	}
 
 	for _, diff := range diffs {
 		switch diff.Kind {
 		case kubernetes.RuleGroupDiffKindAdd:
-			err := c.mimirClient.CreateRuleGroup(ctx, namespace, diff.Desired)
+			err := e.mimirClient.CreateRuleGroup(ctx, namespace, diff.Desired)
 			if err != nil {
 				return err
 			}
-			level.Info(c.log).Log("msg", "added rule group", "namespace", namespace, "group", diff.Desired.Name)
+			level.Info(e.logger).Log("msg", "added rule group", "namespace", namespace, "group", diff.Desired.Name)
 		case kubernetes.RuleGroupDiffKindRemove:
-			err := c.mimirClient.DeleteRuleGroup(ctx, namespace, diff.Actual.Name)
+			err := e.mimirClient.DeleteRuleGroup(ctx, namespace, diff.Actual.Name)
 			if err != nil {
 				return err
 			}
-			level.Info(c.log).Log("msg", "removed rule group", "namespace", namespace, "group", diff.Actual.Name)
+			level.Info(e.logger).Log("msg", "removed rule group", "namespace", namespace, "group", diff.Actual.Name)
 		case kubernetes.RuleGroupDiffKindUpdate:
-			err := c.mimirClient.CreateRuleGroup(ctx, namespace, diff.Desired)
+			err := e.mimirClient.CreateRuleGroup(ctx, namespace, diff.Desired)
 			if err != nil {
 				return err
 			}
-			level.Info(c.log).Log("msg", "updated rule group", "namespace", namespace, "group", diff.Desired.Name)
+			level.Info(e.logger).Log("msg", "updated rule group", "namespace", namespace, "group", diff.Desired.Name)
 		default:
-			level.Error(c.log).Log("msg", "unknown rule group diff kind", "kind", diff.Kind)
+			level.Error(e.logger).Log("msg", "unknown rule group diff kind", "kind", diff.Kind)
 		}
 	}
 
 	// resync mimir state after applying changes
-	return c.syncMimir(ctx)
+	return e.syncMimir(ctx)
+}
+
+// getMimirState returns the cached Mimir ruler state, rule groups indexed by Mimir namespace.
+func (e *eventProcessor) getMimirState() kubernetes.RuleGroupsByNamespace {
+	e.currentStateMtx.Lock()
+	defer e.currentStateMtx.Unlock()
```

Comment:
nit: could be an RLock

Reply:
Fixed in 39d7ee98c511772dca751a6c79594ced837a9b03